core/src/SYCL/Kokkos_SYCL_GraphNodeKernel.hpp

```diff
@@ -0,0 +1,157 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SYCL_GRAPHNODEKERNEL_HPP
+#define KOKKOS_SYCL_GRAPHNODEKERNEL_HPP
+
+#include <Kokkos_Graph_fwd.hpp>
+
+#include <impl/Kokkos_GraphImpl.hpp>
+
+#include <Kokkos_Parallel.hpp>
+#include <Kokkos_Parallel_Reduce.hpp>
+#include <Kokkos_PointerOwnership.hpp>
+
+#include <SYCL/Kokkos_SYCL_GraphNode_Impl.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename PolicyType, typename Functor, typename PatternTag,
+          typename... Args>
+class GraphNodeKernelImpl<Kokkos::Experimental::SYCL, PolicyType, Functor,
+                          PatternTag, Args...>
+    : public PatternImplSpecializationFromTag<
+          PatternTag, Functor, PolicyType, Args...,
+          Kokkos::Experimental::SYCL>::type {
+ public:
+  using Policy       = PolicyType;
+  using graph_kernel = GraphNodeKernelImpl;
+  using base_t       = typename PatternImplSpecializationFromTag<
+      PatternTag, Functor, Policy, Args..., Kokkos::Experimental::SYCL>::type;
+
+  // TODO use the name and executionspace
+  template <typename PolicyDeduced, typename... ArgsDeduced>
+  GraphNodeKernelImpl(std::string, Kokkos::Experimental::SYCL const&,
+                      Functor arg_functor, PolicyDeduced&& arg_policy,
+                      ArgsDeduced&&... args)
+      : base_t(std::move(arg_functor), (PolicyDeduced &&) arg_policy,
+               (ArgsDeduced &&) args...) {}
+
+  template <typename PolicyDeduced>
+  GraphNodeKernelImpl(Kokkos::Experimental::SYCL const& exec_space,
+                      Functor arg_functor, PolicyDeduced&& arg_policy)
+      : GraphNodeKernelImpl("", exec_space, std::move(arg_functor),
+                            (PolicyDeduced &&) arg_policy) {}
+
+  void set_sycl_graph_ptr(
+      sycl::ext::oneapi::experimental::command_graph<
+          sycl::ext::oneapi::experimental::graph_state::modifiable>*
+          arg_graph) {
+    m_graph_ptr = arg_graph;
+  }
+
+  void set_sycl_graph_node_ptr(
+      std::optional<sycl::ext::oneapi::experimental::node>* arg_node) {
+    m_graph_node_ptr = arg_node;
+  }
+
+  std::optional<sycl::ext::oneapi::experimental::node>& get_sycl_graph_node()
```

Comment:
Since we are returning an `optional` anyway, why not avoid the UB if `m_graph_node_ptr` has the null pointer value?  For that matter, why not return the `optional` by value?

Reply:
We are modifying the object obtained through this API so we can't return by value.